getty 偶现panic，程序退出

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

-->


**What happened**:
使用dubbo-go的程序会异常退出

**What you expected to happen**:
如果发生异常，能通过recover捕获异常。
**How to reproduce it (as minimally and precisely as possible)**:
偶现的，我也不知道怎么重现问题
**Anything else we need to know?**:
一些错误提示：
``
2021-08-03T07:25:46.724+0100  DEBUG  getty/pool.go:195  add session{session {client:TCP_CLIENT:24:10.0.2.15:34697<->172.31.229.155:20889}, Read Bytes: 0, Write Bytes: 0, Read Pkgs: 0, Write Pkgs: 0}
2021-08-03T07:25:46.724+0100  ERROR  getty/pool.go:157  tcpConn.SetKeepAlivePeriod() = error:%vset tcp 10.0.2.15:43983->172.31.40.48:28089: setsockopt: invalid argument
github.com/apache/dubbo-go/remoting/getty.(*gettyRPCClient).newSession
  /golang/monitor-agent/vendor/github.com/apache/dubbo-go/remoting/getty/pool.go:157
github.com/apache/dubbo-getty.(*client).connect
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/client.go:391
github.com/apache/dubbo-getty.(*client).reConnect
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/client.go:443
github.com/apache/dubbo-getty.(*client).RunEventLoop
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/client.go:421
panic: @rTimeout < 1

goroutine 6331 [running]:
github.com/apache/dubbo-getty.(*gettyConn).SetReadTimeout(0xc000b5e410, 0x0)
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/connection.go:118 +0x65
github.com/apache/dubbo-go/remoting/getty.(*gettyRPCClient).newSession(0xc00017e000, 0x12436d0, 0xc0004569c0, 0x12436d0, 0xc0004569c0)
  /golang/monitor-agent/vendor/github.com/apache/dubbo-go/remoting/getty/pool.go:171 +0x892
github.com/apache/dubbo-getty.(*client).connect(0xc000a61040)
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/client.go:391 +0xa4
github.com/apache/dubbo-getty.(*client).reConnect(0xc000a61040)
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/client.go:443 +0x56
github.com/apache/dubbo-getty.(*client).RunEventLoop(0xc000a61040, 0xc000254410)
  /golang/monitor-agent/vendor/github.com/apache/dubbo-getty/client.go:421 +0x6d
created by github.com/apache/dubbo-go/remoting/getty.newGettyRPCClientConn
  /golang/monitor-agent/vendor/github.com/apache/dubbo-go/remoting/getty/pool.go:82 +0x22b
``
dubbo-go v1.5.6和v3.0.0都有这个问题

--- a/remoting/getty/config.go
+++ b/remoting/getty/config.go
@@ -30,7 +30,7 @@ import (
 )
 
 const (
-	TCP_READ_WRITE_TIMEOUT_MIN_VALUE = time.Second * 1
+	TCPReadWriteTimeoutMinValue = time.Second * 1
 )
 
 type (
@@ -166,12 +166,11 @@ func (c *GettySessionParam) CheckValidity() error {
 		return perrors.WithMessagef(err, "time.ParseDuration(KeepAlivePeriod{%#v})", c.KeepAlivePeriod)
 	}
 
-	var tcpTimeoutMinVal = TCP_READ_WRITE_TIMEOUT_MIN_VALUE
-	if c.tcpReadTimeout, err = parseTimeDurationByRange(c.TcpReadTimeout, &tcpTimeoutMinVal, nil); err != nil {
+	if c.tcpReadTimeout, err = parseTcpTimeoutDuration(c.TcpReadTimeout); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpReadTimeout{%#v})", c.TcpReadTimeout)
 	}
 
-	if c.tcpWriteTimeout, err = parseTimeDurationByRange(c.TcpWriteTimeout, &tcpTimeoutMinVal, nil); err != nil {
+	if c.tcpWriteTimeout, err = parseTcpTimeoutDuration(c.TcpWriteTimeout); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpWriteTimeout{%#v})", c.TcpWriteTimeout)
 	}
 
@@ -182,19 +181,13 @@ func (c *GettySessionParam) CheckValidity() error {
 	return nil
 }
 
-func parseTimeDurationByRange(timeStr string, min *time.Duration, max *time.Duration) (time.Duration, error) {
+func parseTcpTimeoutDuration(timeStr string) (time.Duration, error) {
 	result, err := time.ParseDuration(timeStr)
 	if err != nil {
-		return *min, err
+		return 0, err
 	}
-	if min != nil && max != nil && *min > *max { // swap
-		min, max = max, min
-	}
-	if min != nil && result < *min {
-		result = *min
-	}
-	if max != nil && result > *max {
-		result = *max
+	if result < TCPReadWriteTimeoutMinValue {
+		return TCPReadWriteTimeoutMinValue, nil
 	}
 	return result, nil
 }

--- a/remoting/getty/config.go
+++ b/remoting/getty/config.go
@@ -162,19 +162,41 @@ func (c *GettySessionParam) CheckValidity() error {
 		return perrors.WithMessagef(err, "time.ParseDuration(KeepAlivePeriod{%#v})", c.KeepAlivePeriod)
 	}
 
-	if c.tcpReadTimeout, err = time.ParseDuration(c.TcpReadTimeout); err != nil {
+	var tcpTimeout *time.Duration
+	tcpTimeoutMinVal := time.Second * 1
+	if tcpTimeout, err = parseTimeDurationByRange(c.TcpReadTimeout, &tcpTimeoutMinVal, nil); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpReadTimeout{%#v})", c.TcpReadTimeout)
 	}
+	c.tcpReadTimeout = *tcpTimeout
 
-	if c.tcpWriteTimeout, err = time.ParseDuration(c.TcpWriteTimeout); err != nil {
+	tcpTimeoutMinVal = time.Second * 5
+	if tcpTimeout, err = parseTimeDurationByRange(c.TcpWriteTimeout, &tcpTimeoutMinVal, nil); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpWriteTimeout{%#v})", c.TcpWriteTimeout)
 	}
+	c.tcpWriteTimeout = *tcpTimeout
 
 	if c.waitTimeout, err = time.ParseDuration(c.WaitTimeout); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(WaitTimeout{%#v})", c.WaitTimeout)
 	}
 
 	return nil
+}
+
+func parseTimeDurationByRange(timeStr string, min *time.Duration, max *time.Duration) (*time.Duration, error) {
+	result, err := time.ParseDuration(timeStr)
+	if err != nil {
+		return nil, err
+	}
+	if min != nil && max != nil && *min > *max {
+		min, max = max, min
+	}
+	if min != nil && result < *min {
+		result = *min
+	}
+	if max != nil && result > *max {
+		result = *max
+	}
+	return &result, nil
 }
 
 // CheckValidity confirm client params.

--- a/remoting/getty/config.go
+++ b/remoting/getty/config.go
@@ -29,6 +29,11 @@ import (
 	"dubbo.apache.org/dubbo-go/v3/config"
 )
 
+const (
+	TCP_READ_TIMEOUT_MIN_VALUE  = time.Second * 1
+	TCP_WRITE_TIMEOUT_MIN_VALUE = time.Second * 5
+)
+
 type (
 	// GettySessionParam is session configuration for getty
 	GettySessionParam struct {
@@ -162,18 +167,13 @@ func (c *GettySessionParam) CheckValidity() error {
 		return perrors.WithMessagef(err, "time.ParseDuration(KeepAlivePeriod{%#v})", c.KeepAlivePeriod)
 	}
 
-	var tcpTimeout *time.Duration
-	tcpTimeoutMinVal := time.Second * 1
-	if tcpTimeout, err = parseTimeDurationByRange(c.TcpReadTimeout, &tcpTimeoutMinVal, nil); err != nil {
+	if c.tcpReadTimeout, err = parseTimeDurationByRange(c.TcpReadTimeout, TCP_READ_TIMEOUT_MIN_VALUE, nil); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpReadTimeout{%#v})", c.TcpReadTimeout)
 	}
-	c.tcpReadTimeout = *tcpTimeout
 
-	tcpTimeoutMinVal = time.Second * 5
-	if tcpTimeout, err = parseTimeDurationByRange(c.TcpWriteTimeout, &tcpTimeoutMinVal, nil); err != nil {
+	if c.tcpWriteTimeout, err = parseTimeDurationByRange(c.TcpWriteTimeout, TCP_WRITE_TIMEOUT_MIN_VALUE, nil); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpWriteTimeout{%#v})", c.TcpWriteTimeout)
 	}
-	c.tcpWriteTimeout = *tcpTimeout
 
 	if c.waitTimeout, err = time.ParseDuration(c.WaitTimeout); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(WaitTimeout{%#v})", c.WaitTimeout)
@@ -182,21 +182,21 @@ func (c *GettySessionParam) CheckValidity() error {
 	return nil
 }
 
-func parseTimeDurationByRange(timeStr string, min *time.Duration, max *time.Duration) (*time.Duration, error) {
+func parseTimeDurationByRange(timeStr string, min time.Duration, max *time.Duration) (time.Duration, error) {
 	result, err := time.ParseDuration(timeStr)
 	if err != nil {
-		return nil, err
+		return min, err
 	}
-	if min != nil && max != nil && *min > *max {
-		min, max = max, min
+	if max != nil && min > *max {
+		min, max = *max, &min
 	}
-	if min != nil && result < *min {
-		result = *min
+	if result < min {
+		result = min
 	}
 	if max != nil && result > *max {
 		result = *max
 	}
-	return &result, nil
+	return result, nil
 }
 
 // CheckValidity confirm client params.

--- a/remoting/getty/config.go
+++ b/remoting/getty/config.go
@@ -30,8 +30,7 @@ import (
 )
 
 const (
-	TCP_READ_TIMEOUT_MIN_VALUE  = time.Second * 1
-	TCP_WRITE_TIMEOUT_MIN_VALUE = time.Second * 5
+	TCP_READ_WRITE_TIMEOUT_MIN_VALUE = time.Second * 1
 )
 
 type (
@@ -167,11 +166,12 @@ func (c *GettySessionParam) CheckValidity() error {
 		return perrors.WithMessagef(err, "time.ParseDuration(KeepAlivePeriod{%#v})", c.KeepAlivePeriod)
 	}
 
-	if c.tcpReadTimeout, err = parseTimeDurationByRange(c.TcpReadTimeout, TCP_READ_TIMEOUT_MIN_VALUE, nil); err != nil {
+	var tcpTimeoutMinVal = TCP_READ_WRITE_TIMEOUT_MIN_VALUE
+	if c.tcpReadTimeout, err = parseTimeDurationByRange(c.TcpReadTimeout, &tcpTimeoutMinVal, nil); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpReadTimeout{%#v})", c.TcpReadTimeout)
 	}
 
-	if c.tcpWriteTimeout, err = parseTimeDurationByRange(c.TcpWriteTimeout, TCP_WRITE_TIMEOUT_MIN_VALUE, nil); err != nil {
+	if c.tcpWriteTimeout, err = parseTimeDurationByRange(c.TcpWriteTimeout, &tcpTimeoutMinVal, nil); err != nil {
 		return perrors.WithMessagef(err, "time.ParseDuration(TcpWriteTimeout{%#v})", c.TcpWriteTimeout)
 	}
 
@@ -182,16 +182,16 @@ func (c *GettySessionParam) CheckValidity() error {
 	return nil
 }
 
-func parseTimeDurationByRange(timeStr string, min time.Duration, max *time.Duration) (time.Duration, error) {
+func parseTimeDurationByRange(timeStr string, min *time.Duration, max *time.Duration) (time.Duration, error) {
 	result, err := time.ParseDuration(timeStr)
 	if err != nil {
-		return min, err
+		return *min, err
 	}
-	if max != nil && min > *max {
-		min, max = *max, &min
+	if min != nil && max != nil && *min > *max { // swap
+		min, max = max, min
 	}
-	if result < min {
-		result = min
+	if min != nil && result < *min {
+		result = *min
 	}
 	if max != nil && result > *max {
 		result = *max

